### PR TITLE
release-22.2: tree: Improve performance of tree.AsJSON

### PR DESCRIPTION
**Pull request by miretskiy**
Backport 2/2 commits from #87968.

/cc @cockroachdb/release

---

Improve performance of `tree.AsJSON` method.

These improvements are important for any query that produces
large number of JSON objects, as well as to changefeeds, which
rely on this function when producing JSON encoded feed.

Most of the changes revolved around modifying underlying types
(s.a. date/timestamp types, box2d, etc) to favor using functions
that append to bytes buffer, instead of relying on slower
functions, such as `fmt.Sprintf`.  The conversion
performance improved around 5-10% for most of the types, and
as high as 50% for time types:

```
Benchmark            old t/op      new t/op    delta
AsJSON/box2d-10    578ns ± 3%    414ns ± 2%   -28.49%  (p=0.000 n=10+9)
AsJSON/box2d[]-10  1.64µs ± 3%   1.19µs ± 4%  -27.14%  (p=0.000 n=10+10)
AsJSON/time-10     232ns ± 2%    103ns ± 1%   -55.61%  (p=0.000 n=10+10)
AsJSON/time[]-10   687ns ± 4%    342ns ± 4%   -50.17%  (p=0.000 n=10+10)
```

Note: Some types in the local benchmark show slight slow down in speed.
No changes were made in those types, and in general, the encoding speed
of these types might be too fast to reliable detect changes:
```
Benchmark            old t/op      new t/op       delta
AsJSON/bool[]-10    65.9ns ± 1%   67.7ns ± 2%    +2.79%  (p=0.001 n=8+9)
```
The emphasis was also placed on reducing allocations.
By relying more heavily on a pooled FmtCtx, which contains
bytes buffer, some conversions resulted in amortized
elimination of allocations (time):
```
Benchmark               old B/op      new t/op    delta
AsJSON/timestamp-10    42.1B ± 3%      0.0B      -100.00%  (p=0.000 n=10+10)
AsJSON/timestamp[]-10  174B ± 4%      60B ± 1%   -65.75%  (p=0.000 n=10+10)
```

Release Note: None
Release Justification: performance improvement

### CODE REVIEW

**Review by yuzefovich — COMMENTED** [review]
Looks like it was a clean backport, so I didn't review carefully. The code LGTM, but since I'm not a TL, I cannot approve it - you'll need to find a TL and persuade them that the backport of a performance improvement is justifiable in this case.

> Reviewed 2 of 2 files at r1, 14 of 14 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/88092#-:-NCMtq-M7cw3E8R-u_V1:bmbsmbj)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy)

File: pkg/util/uint128/uint128.go [hunk 1]
```diff
@@ -30,6 +30,13 @@ func (u Uint128) GetBytes() []byte {
 	return buf
 }
 
+// AppendBytes appends big-endian byte representation to the
+// buffer and returns the buffer.
+func (u Uint128) AppendBytes(buf []byte) []byte {
```
Comment (mgartner) [inline pkg/util/uint128/uint128.go:35]: I don't see this function used anywhere. Am I missing something?

**Review by mgartner — APPROVED** [review]
Very nice!

**Comment by mgartner** [comment]
I've approved the backport, but I think we should consider waiting until 22.2.1 to release this, given that it's not a critical bug fix and we are trying to wrap up the release. What do you think @miretskiy?

File: pkg/util/uint128/uint128.go (hunk 1 above)
Comment (mgartner) [reply pkg/util/uint128/uint128.go:35]: We can leave this backport as-is and remove this function on master and backport that commit too. Or just remove it here if that's easier. I don't feel strongly.

**Comment by mgartner** [comment]
I've added the do-not-merge label to remind us not to merge this until the release-22.2.0 branch is cut.

**Comment by rafiss** [comment]
The release-22.2.0 branch was cut, so this could be merged. However, there seems to be a data race in the new code, which should probably be fixed before backporting: https://github.com/cockroachdb/cockroach/issues/91403

**Comment by rafiss** [comment]
Nevermind, the code I was thinking of is from https://github.com/cockroachdb/cockroach/pull/89884, not this PR.

So this PR is fine to merge.

**Pull request merged**